src/main/java/org/jenkinsci/plugins/workflow/flow/StashManager.java

```diff
@@ -109,26 +114,36 @@ public static void stash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull
      * @see StashAwareArtifactManager#stash
      */
     @SuppressFBWarnings(value="RV_RETURN_VALUE_IGNORED_BAD_PRACTICE", justification="fine if mkdirs returns false")
-    public static void stash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull EnvVars env, @Nonnull TaskListener listener,
+    public static List<File> stash(@Nonnull Run<?,?> build, @Nonnull String name, @Nonnull FilePath workspace, @Nonnull Launcher launcher, @Nonnull EnvVars env, @Nonnull TaskListener listener,
                              @CheckForNull String includes, @CheckForNull String excludes, boolean useDefaultExcludes, boolean allowEmpty) throws IOException, InterruptedException {
         Jenkins.checkGoodName(name);
+        List<File> files = new ArrayList<>();
+        DirScanner.Glob scanner = new DirScanner.Glob(Util.fixEmpty(includes) == null ? "**" : includes, excludes, useDefaultExcludes);
+        scanner.scan(new File(workspace.getRemote()), new FileVisitor() {
+            @Override
+            public void visit(File f, String relativePath) throws IOException {
+                files.add(f);
```

Comment:
We have a `TaskListener` right here. Why not print whatever messages right now, in real time, rather than returning metadata and asking the caller to print messages? There is no need for a new API or a patch to `workflow-basic-steps`.

Reply:
I'd say is nicer to print the messages in the workflow-basic-steps plugin, as is its own responsibility.

Reply:
Easier to leave it to stashing/unstashing implementation, which already has a `TaskListener` exactly for printing such messages as may be relevant. In the case of an alternate implementation such as https://github.com/jenkinsci/artifact-manager-s3-plugin/blob/5e6724a0af838ac48e7c0533520b1e3f081ceef3/src/main/java/io/jenkins/plugins/artifact_manager_jclouds/JCloudsArtifactManager.java#L214-L324 it may be appropriate to do things at different times, or from controller vs. agent, or to show more or less detail, or different formats, etc. etc. From the PoV of `workflow-basic-steps`, these are all implementation details.

Reply:
then, it will be printed only if the implementation has it, not the plugin. Maybe the workflow-basic-step could provide a default implementation, that could be changed in the *Artifact Managers*

Reply:
Not sure I understand what you are saying.

I am simply proposing to move all `TaskListener` messaging into the code which actually does tarring or untarring.